hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -684,6 +684,7 @@ public Object narrowProxy(Object proxy, EntityPersister persister, EntityKey key
 
 			// Otherwise, create the narrowed proxy
 			final HibernateProxy narrowedProxy = (HibernateProxy) persister.createProxy( key.getIdentifier(), session );
```

Comment:
@gbadner Would it help if I performed an unsetSession on the old proxy before overwriting it in the cache? If we leave the super-type proxy in the cache, then subsequent direct fetches of the sub-type will never been the right class type.

Prior to 4.2.12/4.3.5 ([this](https://github.com/hibernate/hibernate-orm/commit/c5bdec20b96a1179d70338644d8c2d5f1408001e#diff-77b73b7dce0dd642db34fbeae2f72350L640) commit), hibernate would (effectively) upgrade the proxy, replace the old cached one, and return the proper sub-typed proxy. Unless that's the direction @sebersole wanted to go, it's an unintentional side-effect of the fix put in for HHH-9071.